Migrate CLI functionality to plugin layer subset 2

cpelley opened the pull request:
Second/final chunk of plugin changes required by EPP.

## Tasks

- improver/cli/extract.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/freezing_rain.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/hail_fraction.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/hail_size.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/interpolate_using_difference.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/lightning_from_cape_and_precip.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/nbhood.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/phase_change_level.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/phase_probability.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/snow_splitter.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/vertical_updraught.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/wet_bulb_freezing_level.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.
- improver/cli/wet_bulb_temperature.py
  - [x] Working acceptance testing.
  - [x] Unittesting coverage.

## Issues

- https://github.com/MetOffice/improver_suite/issues/2079
- https://github.com/metoppv/improver/issues/1998
- https://github.com/MetOffice/barebones_workflow/issues/32
- https://github.com/MetOffice/improver_suite/issues/2253

cpelley commented on improver/wind_calculations/wind_direction.py:
```diff
@@ -123,14 +126,7 @@ def deg_to_complex(
-
-        # Combine components into a complex number and return.
-        return real + 1j * imag
+        return deg_to_complex(angle_deg, radius)
```
defining these elsewhere to avoid circular imports

cpelley replied:
Well, there are numerous unittests relating to the wind stuff involving the use of these things which are tied to the wind data.  I can remove if you are happy that the tests I have written for these functions are enough??

cpelley commented on improver/wind_calculations/wind_direction.py:
```diff
@@ -151,20 +147,7 @@ def complex_to_deg(complex_in: ndarray) -> ndarray:
-        angle = angle.astype(np.float32)
-
-        return angle
+        return complex_to_deg(complex_in)
```
defining these elsewhere to avoid circular imports

cpelley commented on improver_tests/nbhood/nbhood/test_MetaProcModNeighbourhood.py:
Opted for testing the initialisation for exceptions raised only.  The testing is captured in the CLI acceptance testing layer which can be migrated to here when that is removed (no point in duplication).

cpelley commented on improver_tests/psychrometric_calculations/wet_bulb_temperature/test_MetaWetBulbFreezingLevel.py:
The testing is captured in the CLI acceptance testing layer which can be migrated to here when that is removed (no point in duplication).

cpelley commented on improver_tests/psychrometric_calculations/wet_bulb_temperature/test_WetBulbTemperature.py:
```diff
@@ -214,11 +233,11 @@ def test_too_many_cubes(self):
 
-    def test_empty_cube_list(self):
-        """Tests that an error is raised if there is an empty list."""
+    def test_too_few_cubes(self):
```
no cube error handling by as_cubelist function, not this class (i.e. this checks extraction logic of this cubelist by this class).

cpelley commented on improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py:
```diff
@@ -908,11 +908,13 @@ def test_too_many_cubes(self):
             )
 
-    def test_empty_cube_list(self):
+    def test_too_few_cubes(self):
```
no cube error handling by as_cubelist function, not this class (i.e. this checks extraction logic of this cubelist by this class).

cpelley commented:
Finally got this ready for review.  This covers the second and final chunk of IMPROVER CLI/plugin changes required specifically for EPP.
Cheers

bayliffe commented on improver/precipitation_type/freezing_rain.py:
```diff
@@ -268,7 +269,7 @@ def _calculate_freezing_rain_probability(
         return freezing_rain
 
-    def process(self, input_cubes: CubeList) -> Cube:
+    def process(self, *input_cubes: Union[Cube, CubeList]) -> Cube:
```
Technically this can parse a cube, but a cube cannot contain the three diagnostics required, so the plugin will fail. Should we be listing what can technically be handled, or what is actually needed in this typing statement?

cpelley replied:
This is the type of each positional argument (note the asterisk).  That is, each positional argument can be either a cube or cubelist.  The documentation should then stipulate that we expect 3 cubes amongst whatever form of input is provided.
I'll flesh out the documentation here 👍

cpelley replied:
actually the docs do already detail what cubes are expected 🤷

bayliffe replied:
Yeah, you and I clearly think differently, which is why I think you've not followed me here.
I'm saying that the only valid (as in, will actually make the code work) input to this process method is a cubelist. You can technically provide a cube, but you are guaranteed that this plugin will fail as it will not find the necessary inputs.

So my question was whether this typing information should cover all the things that can be passed in (cube, cubelist, str, donkey) or only what can be passed in that will lead to a successful output (cubelist). I'd hedge to the latter, but that's because I think of what the code is trying to do, rather than what it can technically handle, which is why I was asking for your view of it.

cpelley replied:
Ah I see.  In that case, yes my view has been that the type should be what it will accept as input rather than being a reflection of what internal type coercion might be occurring internally.  Type hints also aid IDEs.

That being said, it is a balance between strictness and practicality.  It's often clearer to specify only the types we explicitly expect to handle (we needn't specify every possible type under the sun).

bayliffe commented on improver/precipitation_type/snow_splitter.py:
```diff
@@ -63,7 +64,7 @@ def separate_input_cubes(cubes: CubeList) -> Tuple[Cube, Cube, Cube]:
         return (rain_cube, snow_cube, precip_cube)
 
-    def process(self, cubes: CubeList,) -> Cube:
+    def process(self, *cubes: Union[Cube, CubeList]) -> Cube:
```
Same question about typing, here and anywhere else.

cpelley replied:
we need docs here too :)

cpelley replied:
as above, there is docs here already

bayliffe reviewed (changes requested):
Carwyn, not much to say about this, it largely looks good, just a few queries / requests.

cpelley commented:
@bayliffe, directly relevant change in reflection of this work in https://github.com/MetOffice/dagrunner/pull/40 (you need not review that, I mention only out of interest).

cpelley commented on improver_tests/utilities/complex_conversion/test_complex_to_deg.py:
```diff
@@ -0,0 +1,34 @@
+    """Tests that array of complex values are converted to degrees."""
+    result = complex_to_deg(COMPLEX_ANGLES)
+    assert isinstance(result, np.ndarray)
+    assert_array_almost_equal(result, np.arange(0.0, 360, 10))
```
wind does round trip testing between complex to deg and in reverse with a range of data.  I can remove these if you are happy yes (I was being cautious along with strict timeboxing :P).

cpelley commented on improver/cli/extract.py:
```diff
@@ -52,13 +52,17 @@ def process(
+        ):
+            return cube
+        else:
+            raise
```
I assumed that `ignore_failure` was used for debugging purposes when executed manually hence why I left out of the plugin.  I couldn't see any reference to it being within our configurations at least.  That is I left here as a placeholder for assumed removal.

Happily incorporate into `ExtractSubCube` is this behaviour is to be retained.

cpelley commented:
Changes based on feedback from @bayliffe:
- Moved ignore_error into the plugin of from `improver/cli/extract.py`.
  - Changes to `improver_tests/utilities/cube_extraction/test_ExtractSubCube.py` to reflect this change in behaviour to the plugin.
- Recommended doc change to `improver.utilities.complex_conversion.deg_to_complex`.
- Removed static method `deg_to_complex` and `complex_to_deg` from `WindDirection`.
- Migrated some testing from `improver_tests/wind_calculations/wind_direction/test_WindDirection.py
` to new file `improver_tests/utilities/complex_conversion/test_integration.py`
  - Changes to `improver_tests/wind_calculations/wind_direction/test_WindDirection.py` to ensure successful running of tests that remain.

The only thing I haven't done as I'm not quite following is what you percieve to be an issue in regards to https://github.com/metoppv/improver/pull/2003#discussion_r1665628061
However, I responded that I think it was missed that `*args` is being passed to the function.  The idea is just 1 parameter could be passed if its a cubelist, or 2 or 3 etc.

cpelley commented:
You happy to approve @bayliffe?

cheers

cpelley commented:
I suspect the CI failures with Sphinx are unrelated to this PR (a result of changes in the conda environment).

cpelley commented:
I have merged in master into this branch and changed the target (since subset 1 was merged I forgot to change it here).

cpelley commented:
@mo-robert-purvis, you mind approving once more (it removed your approval after I changed the target to master).
cheers

bayliffe commented on improver/nbhood/nbhood.py:
```diff
@@ -719,3 +722,140 @@ def make_percentile_cube(self, cube: Cube) -> Cube:
+        self,
+        neighbourhood_output: str,
+        radii: List[float],
+        lead_times: List[int],
```
You're missing the way this is handled here: https://github.com/metoppv/improver/blob/31500cb2a2b18b4071ac5bf121fae3f5ac72a0e1/improver/nbhood/__init__.py#L10

cpelley commented:
Thanks both.  @bayliffe, I have made the leadtimes optional once more.  Thanks for double checking this for me.

bayliffe reviewed (approved):
Thanks Carwyn.
